app/apollo/resolvers/group.js

```diff
@@ -485,42 +554,52 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, groupUuids, clusterId }, `${queryName} validating`);
-
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context);
+        logger.info({req_id, user, org_id, clusterId, groupUuids}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         validateString( 'clusterId', clusterId );
         groupUuids.forEach( value => validateString( 'groupUuids', value ) );
 
-        const groups = await models.Group.find({ org_id, uuid: { $in: groupUuids } });
+        const groups = await getAllowedResources(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, null, groupUuids);
```

Comment:
Putting this review comment here for lack of a better location -- it is not directly associated with this line specifically.

While the new code checks that the user is authorized to the input groups (`groupUuids` variable), it does not verify the user is authorized to the groups that will be updated to **remove** the cluster implicitly.   Unlike the other mutations for clusters/groups in this file, the database operation used here does not use not `$pull` or `$push`.  The current implementation here is a straight `$set` -- it can/will remove original groups from the cluster record that are not in the new set of groups for the cluster.

To implement this correctly, more will be needed, perhaps:
- Once the cluster is retrieved and validated (below), find existing groups from the cluster record (i.e. identify which groups it is already a member of)
  - Identify which existing-groups-not-in-the-input the user can READ but not MANAGE -- if any, throw an error like "You are not authorized to modify group XYZ"
  - Identify which existing-groups-not-in-the-input the user is not allowed to even READ -- add those to the list of groups being `$set` so that this operation does not remove them incorrectly

or
- Once the cluster is retrieved and validated (below), find existing groups from the cluster record (i.e. identify which groups it is already a member of)
  - Identify which existing-groups-not-in-the-input the user can READ but not MANAGE -- if any, throw an error like "You are not authorized to modify group XYZ"
  - Identify which existing-groups-not-in-the-input the user is allowed to MANAGE -- add those to a new list of `groupObjsToRemove`
  - Convert the db operation from `$set` to instead `$push` the input groups and simultaneously `$pull` the `groupObjsToRemove`

or
[TBD -- there could be other options]

For either option, ensure the groups actually **removed** from the clusters are passed to pluginContext (this was also a gap in the original pluginContext code -- it would inform plugins what the new set of groups for a cluster is, but wouldn't tell which were removed as a result.

A unit test for trying to set cluster groups when the user does/does not have permissions on all the existing groups would be appropriate, to prevent regression.

Reply:
I suspect from a purely technical perspective, the second option above is superior (avoids potential problems with simultaneous writes from separate requests overwriting changes.... I think)

Reply:
Resolved